Implement click-drag interface for minefields.

tovl opened the pull request:
Fixes #4380.

This reworks the place-minefield order to use a click and drag interface similar to the one for direction airstrikes. This solves a part of #16669 because it is no longer a separate order generator and is now consistent with other unit actions (the other part being attack-move/guard-move).

This is implemented by building the necessary hooks for click-drag orders into the generic unit order code. In this way it can be easily reused for other advanced unit controls such as #4525, #4813 or C&C3 style formations.

matjaeck commented:
There's a crash when you try to abort a dragged overlay with left click (with standard mouse setup).

<details>

```
Exception of type `System.ArgumentNullException`: Value cannot be null.
Parameter name: source
  at System.Linq.Enumerable.Select[TSource,TResult] (System.Collections.Generic.IEnumerable`1[T] source, System.Func`2[T,TResult] selector) [0x0000d] in <6f6cfe3dc7984568b6f7767c7378b138>:0 
  at OpenRA.Orders.UnitOrderGenerator+<Order>d__4.MoveNext () [0x0002e] in <d4ace34ace23427fa5af1ae5a490601a>:0 
  at System.Collections.Generic.LargeArrayBuilder`1[T].AddRange (System.Collections.Generic.IEnumerable`1[T] items) [0x0003d] in <6f6cfe3dc7984568b6f7767c7378b138>:0 
  at System.Collections.Generic.EnumerableHelpers.ToArray[T] (System.Collections.Generic.IEnumerable`1[T] source) [0x00033] in <6f6cfe3dc7984568b6f7767c7378b138>:0 
  at System.Linq.Enumerable.ToArray[TSource] (System.Collections.Generic.IEnumerable`1[T] source) [0x00018] in <6f6cfe3dc7984568b6f7767c7378b138>:0 
  at OpenRA.Mods.Common.Widgets.WorldInteractionControllerWidget.ApplyOrders (OpenRA.World world, OpenRA.MouseInput mi) [0x00062] in <04c6532ba26e45e0a5d23f88b3f0bf5c>:0 
  at OpenRA.Mods.Common.Widgets.WorldInteractionControllerWidget.HandleMouseInput (OpenRA.MouseInput mi) [0x003d8] in <04c6532ba26e45e0a5d23f88b3f0bf5c>:0 
  at OpenRA.Widgets.Widget.HandleMouseInputOuter (OpenRA.MouseInput mi) [0x000a0] in <d4ace34ace23427fa5af1ae5a490601a>:0 
  at OpenRA.Widgets.Widget.HandleMouseInputOuter (OpenRA.MouseInput mi) [0x00052] in <d4ace34ace23427fa5af1ae5a490601a>:0 
  at OpenRA.Widgets.Widget.HandleMouseInputOuter (OpenRA.MouseInput mi) [0x00052] in <d4ace34ace23427fa5af1ae5a490601a>:0 
  at OpenRA.Widgets.Widget.HandleMouseInputOuter (OpenRA.MouseInput mi) [0x00052] in <d4ace34ace23427fa5af1ae5a490601a>:0 
  at OpenRA.Widgets.Ui.HandleInput (OpenRA.MouseInput mi) [0x00030] in <d4ace34ace23427fa5af1ae5a490601a>:0 
  at OpenRA.DefaultInputHandler+<>c__DisplayClass5_0.<OnMouseInput>b__0 () [0x00000] in <d4ace34ace23427fa5af1ae5a490601a>:0 
  at OpenRA.Sync.RunUnsynced[T] (System.Boolean checkSyncHash, OpenRA.World world, System.Func`1[TResult] fn) [0x00023] in <d4ace34ace23427fa5af1ae5a490601a>:0 
  at OpenRA.DefaultInputHandler.OnMouseInput (OpenRA.MouseInput input) [0x0000d] in <d4ace34ace23427fa5af1ae5a490601a>:0 
  at OpenRA.Platforms.Default.Sdl2Input.PumpInput (OpenRA.Platforms.Default.Sdl2PlatformWindow device, OpenRA.IInputHandler inputHandler, System.Nullable`1[T] lockedMousePosition) [0x00227] in <8efc5820dd9844939cf49ae19931d2c9>:0 
  at OpenRA.Platforms.Default.Sdl2PlatformWindow.PumpInput (OpenRA.IInputHandler inputHandler) [0x00006] in <8efc5820dd9844939cf49ae19931d2c9>:0 
  at OpenRA.Renderer.EndFrame (OpenRA.IInputHandler inputHandler) [0x0006c] in <d4ace34ace23427fa5af1ae5a490601a>:0 
  at OpenRA.Game.RenderTick () [0x00188] in <d4ace34ace23427fa5af1ae5a490601a>:0 
  at OpenRA.Game.Loop () [0x0014d] in <d4ace34ace23427fa5af1ae5a490601a>:0 
  at OpenRA.Game.Run () [0x0003c] in <d4ace34ace23427fa5af1ae5a490601a>:0 
  at OpenRA.Game.InitializeAndRun (System.String[] args) [0x00010] in <d4ace34ace23427fa5af1ae5a490601a>:0 
  at OpenRA.Program.Main (System.String[] args) [0x00044] in <d4ace34ace23427fa5af1ae5a490601a>:0

tovl commented:
Fixed the crash.

matjaeck commented:
This has in my opinion some drawbacks:

- The new interface won't work with classic mouse orders. Although there are some other mechanics already that won't work with left click orders (force moving aircraft above other actors), removing this feature could be controversial.
- The old interface allowed you to cancel the overlay without deselecting the unit. This is the most relevant point for me. An option would be to switch to the "can-not-move" cursor when ctrl is released while right mouse button is still held, and then ignore the order when releasing the right mouse button. Currently, although showing the move cursor, the order to lay mines is given in this situation.
- You can draw a selection box while the overlay is still active if you release ctrl and hold left click, see below. This is probably expected but looks weird though.

<details>

![Peek 2019-10-20 22-53](https://user-images.githubusercontent.com/37185497/67166150-12088f80-f38d-11e9-95b9-d0665e100c0b.gif)

</details>

So from a player perspective I'm not entirely convinced to replace the old controls with the proposed interface, probably also because I got used to the old one.

pchote commented:
> The new interface won't work with classic mouse orders.

We'll need to fix at least this one. Can you please file a separate issue for the force-move VTOL issue?

tovl commented:
C&C3 required players to hold both mouse buttons to activate drag-orders. This can be used to allow drag-orders in combination with left-click mouse orders without limiting it to only work with keyboard modifier commands. I would only want to require that with left-click orders though, because for right-click orders it would just be an unnecessary ergonomic downgrade.

tovl commented:
> So from a player perspective I'm not entirely convinced to replace the old controls with the proposed interface, probably also because I got used to the old one.

I am aware that in isolation this would be a more or less neutral change. The important point for me, is not so much about the minelayer itself, but the possibilities that the introduction of this kind of interface opens for more advanced unit controls further down the line. I have deliberately started with the minelayer before introducing anything new however, exactly because I want to first work out the kinks with the interface (such as the three issues you mention).

tovl commented:
Updated:
- The drag order can now be cancelled by left clicking without it immediately deselecting the unit.
- The selection box and drag interface can no longer be active at the same time.
- The cursor is fixed once dragging begins.

Also note that a simple click without dragging is functionally equivalent to what is suggested in https://github.com/OpenRA/OpenRA/issues/16669#issuecomment-500228549.

Implementing drag-orders using both mouse buttons to appease 'classic' controls appears to be way more involved than I had hoped. It seems OpenRA treats both mouse buttons pressed simultaneously as a middle mouse button press somehow. Since that is already reserved for map scrolling (as it should), this is just not going to work.

pchote commented:
Its going to be a while before I can look at this in detail, but here are some thoughts to start with based on a quick test:
* The general idea here seems like a win compared to the current behaviour.
* The white mine-placement overlay sometimes blinks if you do a normal force-target (mouse down, hold just long enough to render one tick, then back up), which looks odd. IMO it would be more polished if we only show that overlay if there are two or more cells.
* I can trigger crashes while trying to test your comment about both mouse buttons acting like middle click (fwiw that isn't the case for me on macOS - could this be caused by your OS's window manager?)
   ```
	Exception of type `System.ArgumentNullException`: Value cannot be null.
	Parameter name: source
	  at System.Linq.Enumerable.Select[TSource,TResult] (System.Collections.Generic.IEnumerable`1[T] source, System.Func`2[T,TResult] selector) [0x00099] in /Users/builder/jenkins/workspace/build-package-osx-mono/2019-06/external/bockbuild/builds/mono-x64/external/corefx/src/System.Linq/src/System/Linq/Select.cs:55 
	  at OpenRA.Orders.UnitOrderGenerator+<Order>d__7.MoveNext () [0x0002e] in /Users/paul/src/OpenRA/OpenRA.Game/Orders/UnitOrderGenerator.cs:61 
	  at System.Collections.Generic.LargeArrayBuilder`1[T].AddRange (System.Collections.Generic.IEnumerable`1[T] items) [0x0003d] in /Users/builder/jenkins/workspace/build-package-osx-mono/2019-06/external/bockbuild/builds/mono-x64/external/corefx/src/Common/src/System/Collections/Generic/LargeArrayBuilder.cs:172 
	  at System.Collections.Generic.EnumerableHelpers.ToArray[T] (System.Collections.Generic.IEnumerable`1[T] source) [0x00033] in /Users/builder/jenkins/workspace/build-package-osx-mono/2019-06/external/bockbuild/builds/mono-x64/external/corefx/src/Common/src/System/Collections/Generic/EnumerableHelpers.cs:84 
	  at System.Linq.Enumerable.ToArray[TSource] (System.Collections.Generic.IEnumerable`1[T] source) [0x0000e] in /Users/builder/jenkins/workspace/build-package-osx-mono/2019-06/external/bockbuild/builds/mono-x64/external/corefx/src/System.Linq/src/System/Linq/ToCollection.cs:18 
	  at OpenRA.Mods.Common.Widgets.WorldInteractionControllerWidget.ApplyOrders (OpenRA.World world, OpenRA.MouseInput mi) [0x00049] in /Users/paul/src/OpenRA/OpenRA.Mods.Common/Widgets/WorldInteractionControllerWidget.cs:212 
	  at OpenRA.Mods.Common.Widgets.WorldInteractionControllerWidget.HandleMouseInput (OpenRA.MouseInput mi) [0x003fe] in /Users/paul/src/OpenRA/OpenRA.Mods.Common/Widgets/WorldInteractionControllerWidget.cs:196 
	  at OpenRA.Widgets.Widget.HandleMouseInputOuter (OpenRA.MouseInput mi) [0x000a0] in /Users/paul/src/OpenRA/OpenRA.Game/Widgets/Widget.cs:400 
	  at OpenRA.Widgets.Widget.HandleMouseInputOuter (OpenRA.MouseInput mi) [0x00052] in /Users/paul/src/OpenRA/OpenRA.Game/Widgets/Widget.cs:391 
	  at OpenRA.Widgets.Widget.HandleMouseInputOuter (OpenRA.MouseInput mi) [0x00052] in /Users/paul/src/OpenRA/OpenRA.Game/Widgets/Widget.cs:391 
	  at OpenRA.Widgets.Widget.HandleMouseInputOuter (OpenRA.MouseInput mi) [0x00052] in /Users/paul/src/OpenRA/OpenRA.Game/Widgets/Widget.cs:391 
	  at OpenRA.Widgets.Ui.HandleInput (OpenRA.MouseInput mi) [0x0002d] in /Users/paul/src/OpenRA/OpenRA.Game/Widgets/Widget.cs:105 
	  at OpenRA.DefaultInputHandler+<>c__DisplayClass5_0.<OnMouseInput>b__0 () [0x00000] in /Users/paul/src/OpenRA/OpenRA.Game/Input/InputHandler.cs:50 
	  at OpenRA.Sync.RunUnsynced[T] (System.Boolean checkSyncHash, OpenRA.World world, System.Func`1[TResult] fn) [0x00023] in /Users/paul/src/OpenRA/OpenRA.Game/Sync.cs:182 
	  at OpenRA.DefaultInputHandler.OnMouseInput (OpenRA.MouseInput input) [0x0000d] in /Users/paul/src/OpenRA/OpenRA.Game/Input/InputHandler.cs:50 
	  at OpenRA.Platforms.Default.Sdl2Input.PumpInput (OpenRA.Platforms.Default.Sdl2PlatformWindow device, OpenRA.IInputHandler inputHandler, System.Nullable`1[T] lockedMousePosition) [0x00206] in /Users/paul/src/OpenRA/OpenRA.Platforms.Default/Sdl2Input.cs:124 
	  at OpenRA.Platforms.Default.Sdl2PlatformWindow.PumpInput (OpenRA.IInputHandler inputHandler) [0x00006] in /Users/paul/src/OpenRA/OpenRA.Platforms.Default/Sdl2PlatformWindow.cs:369 
	  at OpenRA.Renderer.EndFrame (OpenRA.IInputHandler inputHandler) [0x00066] in /Users/paul/src/OpenRA/OpenRA.Game/Renderer.cs:208 
	  at OpenRA.Game.RenderTick () [0x00174] in /Users/paul/src/OpenRA/OpenRA.Game/Game.cs:712 
	  at OpenRA.Game.Loop () [0x0014d] in /Users/paul/src/OpenRA/OpenRA.Game/Game.cs:826 
	  at OpenRA.Game.Run () [0x0003c] in /Users/paul/src/OpenRA/OpenRA.Game/Game.cs:844 
	  at OpenRA.Game.InitializeAndRun (System.String[] args) [0x00010] in /Users/paul/src/OpenRA/OpenRA.Game/Game.cs:262 
	  at OpenRA.Program.Main (System.String[] args) [0x00044] in /Users/paul/src/OpenRA/OpenRA.Game/Support/Program.cs:37 
   ```
   * It doesn't seem possible to target multiple cells in left click order mode, or am I missing something?
   * Can we please split this into multiple commits, first refactoring the order system (possibly over multiple commits) then implementing the new behaviour? The current single commit seems to be doing too many different things at once.
   * I would really like to find a way to not have to explicitly pass `CPos extraLoc` into every `IssueOrder`.

tovl commented:
> It doesn't seem possible to target multiple cells in left click order mode, or am I missing something?

Click-dragging for left-click order mode isn't implemented yet, because of the issues with using both mouse buttons simultaneously. Just using left mouse button drag won't work, because that conflicts with drag selecting.

tovl commented:
Updated:
- I already had a slight time delay before activating the drag overlay that would prevent the overlay blinking when doing a normal click. I now also added a small deadzone. (It is half a cell, because immediately going from zero to two or more tiles also looks odd.)
- Fixed a few crashes.
- Simultaneously pressing both mouse buttons acting like middle click was indeed a "feature" in Cinnamon that was on by default for some reason. With that out of the way, I have now implemented drag orders (using both mouse buttons) for classic mouse style as well.

> I would really like to find a way to not have to explicitly pass CPos extraLoc into every IssueOrder.

I don't see the problem here. That would just make things needlessly convoluted. Besides, we already have implementations of `IssueOrder` that ignore `target` as well, so this fits in the existing pattern. It is really up to the `IssueOrder` implementation to determine what part of the provided information it actually needs. `UnitOrderGenerator` doesn't have any knowledge about that.

matjaeck commented:
> - The drag order can now be cancelled by left clicking without it immediately deselecting the unit.
> - The selection box and drag interface can no longer be active at the same time.

This works well for RMB orders

> - The cursor is fixed once dragging begins.

The move cursor is shown when you release ctrl after clicking but before dragging. Since it is not needed to hold down the ctrl key, could this be implemented more like guard or attack-move or would the following contradict your efforts to improve the flexibility of the feature?

Proposed input:

RMB:
- press ctrl to activate the action cursor
- hold rmb to start dragging, issue order on releasing rmb
- press lmb to cancel action cursor and cancel drag field

LMB:
- press ctrl to activate the action cursor
- hold lmb to start dragging, add an override to avoid drag selection when action cursor is active, issue order on releasing lmb
- press rmb to cancel action cursor and cancel drag field

The points that I don't like too much with the current setup are
- ctrl: what you see is not what you get (move cursor although dragging a minefield)
- lmb: unable to cancel a dragged minefield, order is issued on releasing rmb, the input mechanic is quite an exception and personally I find it inconvenient and error-prone.

I hope my constant nagging doesn't demotivate you.

tovl commented:
> would the following contradict your efforts to improve the flexibility of the feature?

Yes it would. I specifically want to avoid tying this feature to any modifier keys, so that it can be added to any order that might want it regardless of how that order is normally selected. This is also why it is important to have a delay and/or dead zone, to distinguish a drag order from a normal click. Also, making the modifier keys not work like modifiers in this specific instance would run counter to the efforts to make the interface more internally consistent.

I believe I did manage to address you points in an alternative manner though:

> ctrl: what you see is not what you get (move cursor although dragging a minefield)

The order is now only locked in at the same time as the visual overlay appears. If the cursor is changed in any way and released before that, the order should reflect the actually shown cursor type and position. This is consistent with bleed behaviour, where it is the point of release that determines the target for the order and not the point of pressing the button. Order targeters now also explicitly define whether they support dragging or not. If they do not, the order will never get locked in before release of the button.

> lmb: unable to cancel a dragged minefield, order is issued on releasing rmb, the input mechanic is quite an exception and personally I find it inconvenient and error-prone.

I personally found dragging with two buttons at once to be inconvenient in C&C3 as well. (Also IIRC in that game it was impossible to cancel a formation move once you start dragging.)

In any case I came up with an alternative to two button pressing for classic mouse style, that I think is much more elegant. You can now initiate a drag order using just the left mouse button simply by holding for a moment until the visual overlay appears. This then overrides the normal drag selection. You can cancel the drag order by clicking the right mouse button. This is completely consistent with the idea of all orders being given with the left mouse button, and also much more touch screen friendly (which was apparently one of the reasons for keeping the classic one-button interface around). Touch screen users should already be familiar with the concept of a "long press" or "touch and hold".

> I hope my constant nagging doesn't demotivate you.

Not a problem. I expected that this feature would bring with it a lot of edge cases that need to be resolved. I appreciate the feedback.

dragunoff commented:
This conflicts with right mouse scrolling. It's still possible to use it but it's wonky because the scroll gets activated without the order overlay becoming visible. Once the mouse button is released the scrolling stops and the overlay appears.

tovl commented:
I think the best way to fix this is to not allow right mouse scrolling in combination with right mouse orders. Those functions are obviously going to conflict&mdash;no matter what&mdash;without making it more awkward to use in the default setup.

There are at the moment no less than 5 (!) different ways to scroll the map: middle mouse drag, right mouse drag, edge scrolling, cursor keys and mini-map scrolling. I don't think there is any other function in OpenRA that has that much redundant UI. I think sacrificing one of them is an acceptable loss.

dragunoff commented:
Yes, it makes perfect sense to disable scrolling while issuing an order. The player's intention is obviously not to scroll the map.

tovl committed:
Implement click-drag interface for minefields

tovl committed:
Add click-drag interface support to UnitOrderGenerator.

tovl committed:
Rename OrderTargeter references for clarity.

tovl commented:
Rebased.

tovl commented:
The comment was addressed and I have already implemented the suggested solution in a crude manner. I think it still needs a bit of polish to make it clear in the settings menu that right mouse scrolling and right mouse orders cannot be combined. However, I am hesitant to continue with this before someone with more authority than I can confirm that this is an acceptable compromise.

matjaeck commented:
Looks really good at a first glance.

> In any case I came up with an alternative to two button pressing for classic mouse style, that I think is much more elegant. You can now initiate a drag order using just the left mouse button simply by holding for a moment until the visual overlay appears. 

I don't like it too much that this puts classic mouse users in a disadvantage (having to wait for the drag ability) and think it will hide the feature for new players - but it is a good answer to the limitations of left-click orders. What you may want to consider is to show the ability cursor only after the delay when the player is actually able to drag the minefield.

A solution to the controversy about removing right-click scrolling for right-click orders (which currently works, so we'll likely break somebody's workflow) might be to allow changing the drag order mouse button in the settings. People who have been using right mouse scrolling could then set it to middle-mouse.

Codewise I don't like it too much that the `UnitOrderGenerator` does its own input handling and would prefer - if possible - to do all the input handling in the `WorldInteractionControllerWidget` and propagate what's needed to the `UnitOrderGenerator` from there. It also looks like there is an avoidable redundancy in having `isLeftDragging` vs `isRightDragging` vs `IsOrderDragging`.

matjaeck commented on OpenRA.Mods.Cnc/Traits/Minelayer.cs:
```diff
@@ -149,70 +144,51 @@ static IEnumerable<CPos> GetMinefieldCells(CPos start, CPos end, WDist depth)
+			public int OrderPriority { get { return 5; } }
+			public bool TargetOverridesSelection(Actor self, Target target, List<Actor> actorsAt, CPos xy, TargetModifiers modifiers) { return true; }
+			public bool CanDrag { get { return true; } }
+			public bool TargetOverridesSelection(TargetModifiers modifiers) { return true; }
```
This should not be needed.

tovl replied:
Fixed.

pchote commented:
:+1: for restricting right click orders + right click scrolling if this provides a clean solution to the problem. We will need to come up with some ideas for how to represent this in the settings menu before moving forward, though.

tovl committed:
Adapt mouse scroll options in settings menu.

tovl commented:
Updated.

Adapted the settings menu in the following way:
- Instead of having two identical dropdowns dedicated to middle mouse button scrolling and right mouse button scrolling there is now one dropdown for selecting the mouse scroll mode and one for selecting the mouse button used in mouse scrolling.
- It is now also possible to assign the fourth or fifth mouse button for scrolling if you have them.
- When classic mouse orders are disabled, right mouse button will not appear in the dropdown menu.
- When the classic mouse order checkbox is unchecked while the scroll button is set to right mouse button, the dropdown will visibly reset to the default button (middle mouse).

pchote commented:
Would it be too controversial to simplify down to left-click orders always using right-click for scrolling, and right-click orders always using middle-click for scrolling? Our settings menu is a mess as various options grew over several years. I would consider it a win if we could use this opportunity to start simplifying things down to a properly designed set of options.

tovl commented:
> Would it be too controversial to simplify down to left-click orders always using right-click for scrolling, and right-click orders always using middle-click for scrolling?

I had already considered it and, honestly, that would be ideal in my opinion. I just didn't want to push my luck further in taking away options that some people might have become attached to.

I completely agree with you about only supporting a limited but properly designed set of control options instead of the mishmash of edge-cases that make it difficult to fit in new UI elements. This has been my point all along. I am just afraid that it will lead to backlash at some point.

One option I would like to see added at some point though, would be the option to completely swap left and right mouse buttons for all functions in order to support left-handed users. Such a remap would have no adverse influence on the ability to fit in future UI elements because it isn't splitting up or combining functions (like is the case for classic mouse orders or remapping mouse scrolling to a button that is already used for other things).

As for C&C3 (I'm hazy on RA3 because I didn't play it much): It used right mouse button for scrolling and holding both buttons simultaneously for formation move. It didn't differentiate between left-click or right-click orders in this respect at all. I think this is one of the points where C&C3 dropped the ball UI wise; formation moves always felt a bit awkward there. In the mean time middle mouse click-scrolling has become a useful convention in many games and other software and combines really well with right click orders. For left-click orders it makes sense to use the right mouse button for map scrolling, since the right mouse buttons isn't used for much else and is slightly more accessible than the middle mouse button.

matjaeck commented:
What do you think about evaluating the impact of this before doing this? You could add the input settings to the system information and then decide if it's worth it. I'm not affected by the proposal and think it makes sense. But it's polishment for devs (no player cares about messy settings) vs. breaking usability for an unknown amount of players (who will care).

BlackenedRA commented:
> Would it be too controversial to simplify down to left-click orders always using right-click for scrolling, and right-click orders always using middle-click for scrolling? How did C&C3/RA3 handle this?

as a left click orders id be marginally against this as I use middle mouse to scroll. If this change would be more beneficial to the overall project I'd adapt.

dragunoff commented:
RMB was used for scrolling in the gen2 games (TS/RA2) if I remember correctly. But keep in mind that RMB is the only way to deselect in left-click-orders mode.

GraionDilach commented:
Indeed RMB was used in WW C&Cs (both gen1 and gen2) for joystick scrolling and deselection alone. The distinction between click and hold however should be enough to maintain a difference between deselection and scroll. Middle mouse click had no functionality in WW C&Cs.

Another loosely related thing - in OpenRA, you can select through the radar widget, while this was never doable within the WW games, which gives left clicks a new function on the radar screen.

tovl commented:
The idea of restricting both order schemes to using middle mouse for map scrolling has some merit as well. Especially since #17424 further entrenches mouse wheel for zooming, it makes sense to have the middle mouse button be the 'camera' button.

That said, I think this needs more discussion and I don't want that discussion derailing the actual purpose of this PR. I already implemented the least restrictive option and associated settings UI. I would appreciate it if we can defer discussion regarding further restrictions beyond what is strictly necessary to make drag orders work to a follow up and focus on what (if anything) needs to happen to make this mergeable.

tovl commented:
Rebased.

tovl commented:
Because I have to put my OpenRA related activities on hold for a while and the lack of response here, I will park this in 'future' for now.

pchote commented:
I intend to PR a slightly different implementation of the scroll configuration changes here as a prerequisite for fixing #17345 for Next + 1.

![](https://i.imgur.com/v1Xylx3.png)
![](https://i.imgur.com/Rsdxmty.png)

Once that is in, and if review bandwidth allows, I would like to resurrect the rest of this for the milestone too.

tovl commented:
Sadly, #17805 put a spanner in the works here. Is there any chance left now of this ever getting accepted into OpenRA?

pchote commented:
I honestly don't know how to proceed here. I thought the approach was reasonable, but it seems I had massively underestimated the number of people who prefer to use the RMB orders + RMB scroll combination.

tovl commented:
Reopening for discussion.

pchote commented:
IMO it would be better if discussion could happen on #4380 and this PR stay closed until something actionable has been decided. Having pages and pages of unmergable PRs is bad for organization (I have completely lost track of what PRs are active/mergeable/need review) and morale.

tovl commented:
It will not be possible to rebase this without a decision on how to deal with the conflicts introduced by #17805 anyway.
